src/indexer/index_writer.rs

```diff
@@ -342,11 +353,12 @@ impl IndexWriter {
         // dropping the last reference to the segment_updater.
         self.drop_sender();
 
-        let former_workers_handles = std::mem::take(&mut self.workers_join_handle);
-        for join_handle in former_workers_handles {
-            join_handle
-                .join()
-                .map_err(|_| error_in_index_worker_thread("Worker thread panicked."))?
+        let former_workers_future_results = std::mem::take(&mut self.workers_future_results);
+        for future_result in former_workers_future_results.into_iter() {
+            future_result
+                .into_inner()
+                .map_err(|_| TantivyError::Poisoned)?
```

Comment:
When we consume the mutex, we may get an error if the mutex is poisoned, hence the `TantivyError::Poisoned`.

Reply:
There is still something that annoys me with Rayon: before we were able to track if a thread was panicking and raise consequently an `error_in_index_worker_thread`. But it's no more the case in the new version.

Reply:
In case of a panic in the closure, the process that spawned it will be killed. Cf. https://docs.rs/rayon/1.5.1/rayon/struct.ThreadPoolBuilder.html#method.panic_handler

Reply:
> When we consume the mutex, we may get an error if the mutex is poisoned, hence the TantivyError::Poisoned.

I wonder if using the parking_lot crate may be of benefit here? Mutexes are used a fair bit around tantivy (notably around the block cache) although it would be an additional dependency IIRC rayon (along with alot of other libs)  already uses parking_lot so I imagine cargo will be able to re-use.